src/main/java/org/prebid/server/bidder/criteo/CriteoBidder.java

```diff
@@ -0,0 +1,293 @@
+    }
+
+    private static CriteoPublisher buildCriteoPublisher(BidRequest bidRequest, Integer networkId) {
+        final CriteoPublisher.CriteoPublisherBuilder criteoPublisherBuilder = CriteoPublisher.builder();
+        if (networkId != null) {
+            criteoPublisherBuilder.networkId(networkId);
+        }
+
+        if (bidRequest.getApp() != null) {
+            criteoPublisherBuilder.bundleId(bidRequest.getApp().getBundle());
+        }
+
+        if (bidRequest.getSite() != null) {
+            criteoPublisherBuilder
+                    .siteId(bidRequest.getSite().getId())
+                    .url(bidRequest.getSite().getPage());
+        }
+
+        return criteoPublisherBuilder.build();
+    }
+
+    private static CriteoUser buildCriteoUser(BidRequest bidRequest, ExtRegs extRegs) {
+        final CriteoUser.CriteoUserBuilder criteoUserBuilder = CriteoUser.builder();
+        if (bidRequest.getUser() != null) {
+            criteoUserBuilder.cookieId(bidRequest.getUser().getBuyeruid());
+        }
+
+        if (bidRequest.getDevice() != null) {
+            criteoUserBuilder
+                    .deviceIdType(determineDeviceIdType(bidRequest.getDevice().getOs()))
+                    .deviceOs(bidRequest.getDevice().getOs())
+                    .deviceId(bidRequest.getDevice().getIfa())
+                    .ip(bidRequest.getDevice().getIp())
+                    .ipV6(bidRequest.getDevice().getIpv6())
+                    .userAgent(bidRequest.getDevice().getUa());
+        }
+
+        if (extRegs != null) {
+            criteoUserBuilder.uspIab(extRegs.getUsPrivacy());
+        }
+
+        return criteoUserBuilder.build();
+    }
+
+    private static String determineDeviceIdType(String deviceOs) {
+        switch (deviceOs.toLowerCase()) {
+            case "ios":
+                return "idfa";
+            case "android":
+                return "gaid";
+            default:
+                return "unknown";
+        }
+    }
+
+    private static CriteoGdprConsent buildCriteoGdprConsent(BidRequest bidRequest, ExtRegs extRegs) {
+        final CriteoGdprConsent.CriteoGdprConsentBuilder criteoGdprConsentBuilder = CriteoGdprConsent.builder();
+        if (bidRequest.getUser() != null && bidRequest.getUser().getExt() != null) {
+            final ExtUser extUser = bidRequest.getUser().getExt();
+            criteoGdprConsentBuilder.consentData(extUser.getConsent());
+        }
+
+        if (extRegs != null && extRegs.getGdpr() != null) {
+            criteoGdprConsentBuilder.gdprApplies(extRegs.getGdpr() == 1);
+        }
+        return criteoGdprConsentBuilder.build();
+    }
+
+    private static MultiMap resolveHeaders(CriteoRequest criteoRequest) {
+        final MultiMap headers = HttpUtil.headers();
+
+        if (StringUtils.isNotBlank(criteoRequest.getUser().getCookieId())) {
+            headers.add(HttpUtil.COOKIE_HEADER, String.format("uid=%s", criteoRequest.getUser().getCookieId()));
```

Comment:
Please check if it would convenient to use `org.prebid.server.util.HttpUtil#addHeaderIfValueIsNotEmpty`.
Also, we need to be sure `criteoRequest.getUser()` is not null.

Reply:
Done